build_rules.gradle

```diff
@@ -0,0 +1,275 @@
+  apply plugin: "maven"
+  apply plugin: "java"
+
+  sourceCompatibility = configuration.javaVersion
+  targetCompatibility = configuration.javaVersion
+  tasks.withType(JavaCompile) {
+    options.encoding = "UTF-8"
+  }
+
+  test {
+    // Configure the gradle include defaults to be equivalent to the maven-surefire-plugin
+    // http://maven.apache.org/components/surefire/maven-surefire-plugin/test-mojo.html
+    include "**/Test*.class"
+    include "**/*Test.class"
+    include "**/*Tests.class"
+    include "**/*TestCase.class"
+    useJUnit { }
+  }
+
+  apply plugin: "net.ltgt.apt"
+  dependencies {
+    def auto_value = "com.google.auto.value:auto-value:1.5.1"
+    def auto_service = "com.google.auto.service:auto-service:1.0-rc2"
+    compileOnly auto_value
+    apt auto_value
+    testCompileOnly auto_value
+    testApt auto_value
+
+    compileOnly auto_service
+    apt auto_service
+    testCompileOnly auto_service
+    testApt auto_service
+  }
+
+  // Add the optional and provided scopes for dependencies
+  apply plugin: 'propdeps'
+  apply plugin: 'propdeps-maven'
+
+  apply plugin: 'checkstyle'
+
+  tasks.withType(Checkstyle) {
+    configFile = project(":").file("sdks/java/build-tools/src/main/resources/beam/checkstyle.xml")
+    configProperties = [ "checkstyle.suppressions.file": project(":").file("sdks/java/build-tools/src/main/resources/beam/suppressions.xml") ]
+    showViolations = true
+    maxErrors = 0
+  }
+
+  apply plugin: 'com.diffplug.gradle.spotless'
+  spotless {
+    java {
+      target rootProject.fileTree(rootProject.rootDir) {
+        include 'sdks/java/**/*.java'
+      }
+      // Code formatting disabled because style rules are out of date.
+      // eclipse().configFile(rootProject.file('sdks/java/build-tools/src/main/resources/beam/beam-codestyle.xml'))
+    }
+  }
+
+  if (configuration.enableFindbugs) {
+    apply plugin: 'findbugs'
+    findbugs {
+      excludeFilter = rootProject.file('sdks/java/build-tools/src/main/resources/beam/findbugs-filter.xml')
+      sourceSets = [sourceSets.main]
+    }
+  }
+
+  if (configuration.enableShadow) {
+    apply plugin: 'com.github.johnrengelman.shadow'
+
+    // Shade guava in all our dependencies.
+    shadowJar {
+      classifier = "shaded"
+      mergeServiceFiles()
+      dependencies {
+        exclude(".*")
+        include(dependency(library.java.guava))
+      }
+      relocate("com.google.common", "org.apache.beam." + project.name.replace("-", ".") + ".repackaged.com.google.common") {
+        // com.google.common is too generic, need to exclude guava-testlib
+        exclude "com.google.common.collect.testing.**"
+        exclude "com.google.common.escape.testing.**"
+        exclude "com.google.common.testing.**"
+        exclude "com.google.common.util.concurrent.testing.**"
+      }
+    }
+
+    // Create a new configuration 'shadowTest' like 'shadow' for the test scope
+    configurations {
```

Comment:
<!--new_thread; commit:8ca3fdca276a74973cd428d7321b995f7b7ebccf; resolved:0-->
This configuration stuff does make me think it would be ideal to have a single shaded dep in its own module, so there's none of this spooky configuration at a distance of all projects in really nontrivial ways.